aws-amplify/aws-sdk-ios
# fix(AWSLocation): Add AWSLocationXCF target for building frameworks

*Issue #, if available:*

*Description of changes:*
Add a duplicate target of AWSLocation called `AWSLocationXCF` and so that the types added in swift like TrackingOption under the AWSLocation is exposed as AWSLocationXCF in the framework, so it does not conflict with the objective-c AWSLocation interface (`AWSLocationService.h` defines the AWSLocation interface)

*Check points:*


*Testing done*

1. Steps taken to reproduce the issue
```
export project_file=./AWSiOSSDKv2.xcodeproj/
export framework=AWSLocation
export archive_path=./archives
xcodebuild archive -project $project_file -scheme $framework -destination "generic/platform=iOS Simulator" -archivePath $archive_path SKIP_INSTALL=NO BUILD_LIBRARY_FOR_DISTRIBUTION=YES 
xcodebuild -create-xcframework -framework archives.xcarchive/Products/Library/Frameworks/AWSLocation.framework -output archives/AWSLocation.xcframework
```

2. Create new App, Targets -> Build Phases -> Link Binary with Libraries -> add Files -> Add Other -> find and add AWSLocation.xcframework from archives folder

3. Add `import AWSLocation` and try to compile, see issue: 
```
Failed to build module 'AWSLocation' from its module interface; it may have been damaged or it may have triggered a bug in the Swift compiler when it was produced
```
There will be build time errors such as 
```
  public let errorType: AWSLocation.TrackingError.ErrorType
  // 'TrackingError' is not a member type of 'AWSLocation'
```

4. Build AWSLocationXCF instead:
```
export frameworkXCF=AWSLocationXCF
xcodebuild archive -project $project_file -scheme $frameworkXCF -destination "generic/platform=iOS Simulator" -archivePath $archive_path SKIP_INSTALL=NO BUILD_LIBRARY_FOR_DISTRIBUTION=YES 
xcodebuild -create-xcframework -framework archives.xcarchive/Products/Library/Frameworks/AWSLocationXCF.framework -output archives/AWSLocationXCF.xcframework
```

5. Back in the app, remove the previously added framework and add `AWSLocationXCF.framework`

6. Build should still fail on missing dependency frameworks such as AWSCore. Build some of the dependencies and add them in.

7. Update the import to `AWSLocationXCF`

8. Code usage remains the same (https://docs.amplify.aws/guides/location-service/setting-up-your-app/q/platform/ios)

*bump version script testing*
`/usr/local/bin/python3 ./CircleciScripts/bump_sdk_version.py  "/Users/user/aws-amplify/aws-sdk-ios" 2.23.2`
output: https://github.com/aws-amplify/aws-sdk-ios/compare/fix/awslocationxcf...lawmicha/test-bump-version?expand=1

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
